File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraServer.java
package org.emulinker.kaillera.model;

import java.net.InetSocketAddress;
import java.util.Collection;
import org.emulinker.kaillera.access.AccessManager;
import org.emulinker.kaillera.model.event.KailleraEventListener;
import org.emulinker.kaillera.model.exception.*;
import org.emulinker.kaillera.model.impl.KailleraUserImpl;
import org.emulinker.kaillera.model.impl.Trivia;
import org.emulinker.release.*;

public interface KailleraServer {
  public ReleaseInfo getReleaseInfo();

  public int getNumUsers();

  public int getNumGames();

  public int getMaxUsers();

  public void setTrivia(Trivia trivia);

  public void setSwitchTrivia(boolean switchTrivia);

  public Trivia getTrivia();

  public boolean getSwitchTrivia();

  public void announce(String announcement, boolean gamesAlso, KailleraUserImpl user);

  public int getMaxGames();

  public int getMaxPing();

  public AccessManager getAccessManager();

  public Collection<? extends KailleraUser> getUsers();

  public Collection<? extends KailleraGame> getGames();

  public KailleraUser getUser(int userID);

  public KailleraGame getGame(int gameID);

  public boolean checkMe(KailleraUser user, String message);

  public KailleraUser newConnection(
      InetSocketAddress clientSocketAddress, String protocol, KailleraEventListener listener)
      throws ServerFullException, NewConnectionException;

  public void login(KailleraUser user)
      throws PingTimeException, ClientAddressException, ConnectionTypeException, UserNameException,
          LoginException;

  public void chat(KailleraUser user, String message) throws ChatException, FloodException;

  public KailleraGame createGame(KailleraUser user, String romName)
      throws CreateGameException, FloodException;

  public void quit(KailleraUser user, String message)
      throws QuitException, DropGameException, QuitGameException, CloseGameException;

  public abstract void start();

  public abstract void stop();
}

File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraServer.kt
package org.emulinker.kaillera.model

import java.net.InetSocketAddress
import kotlin.Throws
import org.emulinker.kaillera.access.AccessManager
import org.emulinker.kaillera.model.event.KailleraEventListener
import org.emulinker.kaillera.model.exception.*
import org.emulinker.kaillera.model.impl.KailleraGameImpl
import org.emulinker.kaillera.model.impl.KailleraUserImpl
import org.emulinker.kaillera.model.impl.Trivia
import org.emulinker.kaillera.release.ReleaseInfo

interface KailleraServer {
  val maxUsers: Int
  val maxGames: Int
  val maxPing: Int

  val releaseInfo: ReleaseInfo
  val accessManager: AccessManager
  val users: Collection<KailleraUserImpl>
  val games: Collection<KailleraGameImpl>
  var trivia: Trivia?
  var switchTrivia: Boolean

  fun announce(message: String, gamesAlso: Boolean)
  fun announce(message: String, gamesAlso: Boolean, targetUser: KailleraUserImpl?)
  fun getUser(userID: Int): KailleraUser?
  fun getGame(gameID: Int): KailleraGame?
  fun checkMe(user: KailleraUser, message: String): Boolean

  @Throws(ServerFullException::class, NewConnectionException::class)
  fun newConnection(
      clientSocketAddress: InetSocketAddress, protocol: String, listener: KailleraEventListener
  ): KailleraUser

  @Throws(
      PingTimeException::class,
      ClientAddressException::class,
      ConnectionTypeException::class,
      UserNameException::class,
      LoginException::class)
  suspend fun login(user: KailleraUser)

  @Throws(ChatException::class, FloodException::class)
  fun chat(user: KailleraUser, message: String)

  @Throws(CreateGameException::class, FloodException::class)
  suspend fun createGame(user: KailleraUser, romName: String?): KailleraGame?

  @Throws(
      QuitException::class,
      DropGameException::class,
      QuitGameException::class,
      CloseGameException::class)
  fun quit(user: KailleraUser, message: String?)
  suspend fun stop()
}
